# Video: Visualizing Byte Pair Encoding

One key step in the evolution of large language models was transitioning to predicting tokens instead of just bytes.
This video shows how the compression method byte pair encoding was repurposed to identify frequent text patterns for token selection, and gives examples of the increased semantic information in the resulting tokens.


Script:
* Byte pair encoding was originally designed as a compression algorithm.
* Byte pair encoding identifies common sequences of bytes in the data set starting from pairs of adjacent bytes, and builds a table of them to aid in compression.
* This original motivation for compression ties to our language modeling goals.
* Common sequences for compression are also sequences that we want to predict often.
* A few different language modeling papers used byte pair encoding to aid language analysis, and GPT-2 was the first large language model to use them for predicting tokens.

In [ ]:
test_text = '''Article I
Section 1

All legislative Powers herein granted shall be vested in a Congress of the United States, which shall consist of a Senate and House of Representatives.

Section 2

The House of Representatives shall be composed of Members chosen every second Year by the People of the several States, and the Electors in each State shall have the Qualifications requisite for Electors of the most numerous Branch of the State Legislature.

No Person shall be a Representative who shall not have attained to the Age of twenty five Years, and been seven Years a Citizen of the United States, and who shall not, when elected, be an Inhabitant of that State in which he shall be chosen.

Representatives and direct Taxes shall be apportioned among the several States which may be included within this Union, according to their respective Numbers, which shall be determined by adding to the whole Number of free Persons, including those bound to Service for a Term of Years, and excluding Indians not taxed, three fifths of all other Persons. The actual Enumeration shall be made within three Years after the first Meeting of the Congress of the United States, and within every subsequent Term of ten Years, in such Manner as they shall by Law direct. The Number of Representatives shall not exceed one for every thirty Thousand, but each State shall have at Least one Representative; and until such enumeration shall be made, the State of New Hampshire shall be entitled to chuse three, Massachusetts eight, Rhode Island and Providence Plantations one, Connecticut five, New-York six, New Jersey four, Pennsylvania eight, Delaware one, Maryland six, Virginia ten, North Carolina five, South Carolina five, and Georgia three.

When vacancies happen in the Representation from any State, the Executive Authority thereof shall issue Writs of Election to fill such Vacancies.

The House of Representatives shall chuse their Speaker and other Officers; and shall have the sole Power of Impeachment.

Section 3

The Senate of the United States shall be composed of two Senators from each State, chosen by the Legislature thereof, for six Years; and each Senator shall have one Vote.

Immediately after they shall be assembled in Consequence of the first Election, they shall be divided as equally as may be into three Classes. The Seats of the Senators of the first Class shall be vacated at the Expiration of the second Year, of the second Class at the Expiration of the fourth Year, and of the third Class at the Expiration of the sixth Year, so that one third may be chosen every second Year; and if Vacancies happen by Resignation, or otherwise, during the Recess of the Legislature of any State, the Executive thereof may make temporary Appointments until the next Meeting of the Legislature, which shall then fill such Vacancies.

No Person shall be a Senator who shall not have attained to the Age of thirty Years, and been nine Years a Citizen of the United States, and who shall not, when elected, be an Inhabitant of that State for which he shall be chosen.

The Vice President of the United States shall be President of the Senate, but shall have no Vote, unless they be equally divided.

The Senate shall chuse their other Officers, and also a President pro tempore, in the Absence of the Vice President, or when he shall exercise the Office of President of the United States.

The Senate shall have the sole Power to try all Impeachments. When sitting for that Purpose, they shall be on Oath or Affirmation. When the President of the United States is tried, the Chief Justice shall preside: And no Person shall be convicted without the Concurrence of two thirds of the Members present.

Judgment in Cases of Impeachment shall not extend further than to removal from Office, and disqualification to hold and enjoy any Office of honor, Trust or Profit under the United States: but the Party convicted shall nevertheless be liable and subject to Indictment, Trial, Judgment and Punishment, according to Law.

Section 4

The Times, Places and Manner of holding Elections for Senators and Representatives, shall be prescribed in each State by the Legislature thereof; but the Congress may at any time by Law make or alter such Regulations, except as to the Places of chusing Senators.

The Congress shall assemble at least once in every Year, and such Meeting shall be on the first Monday in December, unless they shall by Law appoint a different Day.

Section 5

Each House shall be the Judge of the Elections, Returns and Qualifications of its own Members, and a Majority of each shall constitute a Quorum to do Business; but a smaller Number may adjourn from day to day, and may be authorized to compel the Attendance of absent Members, in such Manner, and under such Penalties as each House may provide.

Each House may determine the Rules of its Proceedings, punish its Members for disorderly Behaviour, and, with the Concurrence of two thirds, expel a Member.

Each House shall keep a Journal of its Proceedings, and from time to time publish the same, excepting such Parts as may in their Judgment require Secrecy; and the Yeas and Nays of the Members of either House on any question shall, at the Desire of one fifth of those Present, be entered on the Journal.

Neither House, during the Session of Congress, shall, without the Consent of the other, adjourn for more than three days, nor to any other Place than that in which the two Houses shall be sitting.

Section 6

The Senators and Representatives shall receive a Compensation for their Services, to be ascertained by Law, and paid out of the Treasury of the United States. They shall in all Cases, except Treason, Felony and Breach of the Peace, be privileged from Arrest during their Attendance at the Session of their respective Houses, and in going to and returning from the same; and for any Speech or Debate in either House, they shall not be questioned in any other Place.

No Senator or Representative shall, during the Time for which he was elected, be appointed to any civil Office under the Authority of the United States, which shall have been created, or the Emoluments whereof shall have been encreased during such time; and no Person holding any Office under the United States, shall be a Member of either House during his Continuance in Office.

Section 7

All Bills for raising Revenue shall originate in the House of Representatives; but the Senate may propose or concur with Amendments as on other Bills.

Every Bill which shall have passed the House of Representatives and the Senate, shall, before it become a Law, be presented to the President of the United States; If he approve he shall sign it, but if not he shall return it, with his Objections to that House in which it shall have originated, who shall enter the Objections at large on their Journal, and proceed to reconsider it. If after such Reconsideration two thirds of that House shall agree to pass the Bill, it shall be sent, together with the Objections, to the other House, by which it shall likewise be reconsidered, and if approved by two thirds of that House, it shall become a Law. But in all such Cases the Votes of both Houses shall be determined by yeas and Nays, and the Names of the Persons voting for and against the Bill shall be entered on the Journal of each House respectively. If any Bill shall not be returned by the President within ten Days (Sundays excepted) after it shall have been presented to him, the Same shall be a Law, in like Manner as if he had signed it, unless the Congress by their Adjournment prevent its Return, in which Case it shall not be a Law.

Every Order, Resolution, or Vote to which the Concurrence of the Senate and House of Representatives may be necessary (except on a question of Adjournment) shall be presented to the President of the United States; and before the Same shall take Effect, shall be approved by him, or being disapproved by him, shall be repassed by two thirds of the Senate and House of Representatives, according to the Rules and Limitations prescribed in the Case of a Bill.

Section 8

The Congress shall have Power To lay and collect Taxes, Duties, Imposts and Excises, to pay the Debts and provide for the common Defence and general Welfare of the United States; but all Duties, Imposts and Excises shall be uniform throughout the United States;

To borrow Money on the credit of the United States;

To regulate Commerce with foreign Nations, and among the several States, and with the Indian Tribes;

To establish an uniform Rule of Naturalization, and uniform Laws on the subject of Bankruptcies throughout the United States;

To coin Money, regulate the Value thereof, and of foreign Coin, and fix the Standard of Weights and Measures;

To provide for the Punishment of counterfeiting the Securities and current Coin of the United States;

To establish Post Offices and post Roads;

To promote the Progress of Science and useful Arts, by securing for limited Times to Authors and Inventors the exclusive Right to their respective Writings and Discoveries;

To constitute Tribunals inferior to the supreme Court;

To define and punish Piracies and Felonies committed on the high Seas, and Offences against the Law of Nations;

To declare War, grant Letters of Marque and Reprisal, and make Rules concerning Captures on Land and Water;

To raise and support Armies, but no Appropriation of Money to that Use shall be for a longer Term than two Years;

To provide and maintain a Navy;

To make Rules for the Government and Regulation of the land and naval Forces;

To provide for calling forth the Militia to execute the Laws of the Union, suppress Insurrections and repel Invasions;

To provide for organizing, arming, and disciplining, the Militia, and for governing such Part of them as may be employed in the Service of the United States, reserving to the States respectively, the Appointment of the Officers, and the Authority of training the Militia according to the discipline prescribed by Congress;

To exercise exclusive Legislation in all Cases whatsoever, over such District (not exceeding ten Miles square) as may, by Cession of particular States, and the Acceptance of Congress, become the Seat of Government of the United States, and to exercise like Authority over all Places purchased by the Consent of the Legislature of the State in which the Same shall be, for the Erection of Forts, Magazines, Arsenals, dock-Yards, and other needful Buildings;–And

To make all Laws which shall be necessary and proper for carrying into Execution the foregoing Powers, and all other Powers vested by this Constitution in the Government of the United States, or in any Department or Officer thereof.

Section 9

The Migration or Importation of such Persons as any of the States now existing shall think proper to admit, shall not be prohibited by the Congress prior to the Year one thousand eight hundred and eight, but a Tax or duty may be imposed on such Importation, not exceeding ten dollars for each Person.

The Privilege of the Writ of Habeas Corpus shall not be suspended, unless when in Cases of Rebellion or Invasion the public Safety may require it.

No Bill of Attainder or ex post facto Law shall be passed.

No Capitation, or other direct, Tax shall be laid, unless in Proportion to the Census or enumeration herein before directed to be taken.

No Tax or Duty shall be laid on Articles exported from any State.

No Preference shall be given by any Regulation of Commerce or Revenue to the Ports of one State over those of another: nor shall Vessels bound to, or from, one State, be obliged to enter, clear, or pay Duties in another.

No Money shall be drawn from the Treasury, but in Consequence of Appropriations made by Law; and a regular Statement and Account of the Receipts and Expenditures of all public Money shall be published from time to time.

No Title of Nobility shall be granted by the United States: And no Person holding any Office of Profit or Trust under them, shall, without the Consent of the Congress, accept of any present, Emolument, Office, or Title, of any kind whatever, from any King, Prince, or foreign State.

Section 10

No State shall enter into any Treaty, Alliance, or Confederation; grant Letters of Marque and Reprisal; coin Money; emit Bills of Credit; make any Thing but gold and silver Coin a Tender in Payment of Debts; pass any Bill of Attainder, ex post facto Law, or Law impairing the Obligation of Contracts, or grant any Title of Nobility.

No State shall, without the Consent of the Congress, lay any Imposts or Duties on Imports or Exports, except what may be absolutely necessary for executing it's inspection Laws: and the net Produce of all Duties and Imposts, laid by any State on Imports or Exports, shall be for the Use of the Treasury of the United States; and all such Laws shall be subject to the Revision and Controul of the Congress.

No State shall, without the Consent of Congress, lay any Duty of Tonnage, keep Troops, or Ships of War in time of Peace, enter into any Agreement or Compact with another State, or with a foreign Power, or engage in War, unless actually invaded, or in such imminent Danger as will not admit of delay.
'''

In [ ]:
import tiktoken

Script:
* First I will show you some tokens picked by tiktoken with the tokenizer for OpenAI's gpt-4o model.

In [ ]:
encoder = tiktoken.encoding_for_model("gpt-4o")

In [ ]:
encoder.n_vocab

200019

Script:
* This encoder is based on 200,019 tokens.
* Let's look at the first few tokens assigned.

In [ ]:
encoder.decode([0])

'!'

Script:
* Token 0 is an exclamation point.

In [ ]:
encoder.decode([1])

'"'

Script:
* Token 1 is a double quote.

In [ ]:
encoder.decode([2])

'#'

Script:
* Token 2 is the sharp or pound sign depending on whether you embrace a music or programming background more.

In [ ]:
encoder.decode([3])

'$'

Script:
* Token 3 is a dollar sign.
* Let's look at more at a time.

In [ ]:
for i in range(256):
    print(i, repr(encoder.decode([i])))

0 '!'
1 '"'
2 '#'
3 '$'
4 '%'
5 '&'
6 "'"
7 '('
8 ')'
9 '*'
10 '+'
11 ','
12 '-'
13 '.'
14 '/'
15 '0'
16 '1'
17 '2'
18 '3'
19 '4'
20 '5'
21 '6'
22 '7'
23 '8'
24 '9'
25 ':'
26 ';'
27 '<'
28 '='
29 '>'
30 '?'
31 '@'
32 'A'
33 'B'
34 'C'
35 'D'
36 'E'
37 'F'
38 'G'
39 'H'
40 'I'
41 'J'
42 'K'
43 'L'
44 'M'
45 'N'
46 'O'
47 'P'
48 'Q'
49 'R'
50 'S'
51 'T'
52 'U'
53 'V'
54 'W'
55 'X'
56 'Y'
57 'Z'
58 '['
59 '\\'
60 ']'
61 '^'
62 '_'
63 '`'
64 'a'
65 'b'
66 'c'
67 'd'
68 'e'
69 'f'
70 'g'
71 'h'
72 'i'
73 'j'
74 'k'
75 'l'
76 'm'
77 'n'
78 'o'
79 'p'
80 'q'
81 'r'
82 's'
83 't'
84 'u'
85 'v'
86 'w'
87 'x'
88 'y'
89 'z'
90 '{'
91 '|'
92 '}'
93 '~'
94 '�'
95 '�'
96 '�'
97 '�'
98 '�'
99 '�'
100 '�'
101 '�'
102 '�'
103 '�'
104 '�'
105 '�'
106 '�'
107 '�'
108 '�'
109 '�'
110 '�'
111 '�'
112 '�'
113 '�'
114 '�'
115 '�'
116 '�'
117 '�'
118 '�'
119 '�'
120 '�'
121 '�'
122 '�'
123 '�'
124 '�'
125 '�'
126 '�'
127 '�'
128 '�'
129 '�'
130 '�'
131 '�'
132 '�'
133 '�'
134 '�'
135 '�'
136 '�'
137 '�'
138 '

Script:
* We saw all the English alphabet and common punctuation symbols.
* We also some symbols that were unprintable, at least in the context of this notebook.

In [ ]:
for i in range(256, 512):
    print(i, repr(encoder.decode([i])))

256 '  '
257 '    '
258 'in'
259 'er'
260 ' t'
261 ' a'
262 'en'
263 'on'
264 're'
265 ' s'
266 'at'
267 'or'
268 'es'
269 '        '
270 'an'
271 '   '
272 ' d'
273 'he'
274 ' c'
275 ' p'
276 'is'
277 'ar'
278 'it'
279 '\n\n'
280 'al'
281 '�'
282 'le'
283 'ou'
284 ' m'
285 ' f'
286 ' w'
287 ' b'
288 'as'
289 'ing'
290 ' the'
291 'ic'
292 'et'
293 ' o'
294 'ion'
295 'ed'
296 'el'
297 ' n'
298 'ro'
299 'ent'
300 ' �'
301 'nd'
302 'st'
303 '�'
304 'а'
305 ' l'
306 ' in'
307 ';\n'
308 'ct'
309 '       '
310 'om'
311 'il'
312 ' h'
313 'am'
314 ' ='
315 'id'
316 ' to'
317 'о'
318 '�'
319 ' e'
320 'ا'
321 'im'
322 ' re'
323 ' v'
324 'ad'
325 ' th'
326 ' and'
327 'е'
328 ' of'
329 ' g'
330 'ur'
331 'и'
332 'ch'
333 ' �'
334 ' de'
335 '\t\t'
336 ' S'
337 ' u'
338 'т'
339 'ut'
340 'ol'
341 'н'
342 ' y'
343 'ig'
344 'se'
345 'р'
346 'ot'
347 'em'
348 'ag'
349 'iv'
350 ' ('
351 'qu'
352 '           '
353 ' T'
354 ' {'
355 ' A'
356 'ay'
357 ' I'
358 '�'
359 'ac'
360 '�'
361 'ul'
362 ');\n'
363 ' C

Script:
* In this batch, we saw many pairs of letters glommed together.
* There were also characters from languages besides English.
* I saw a number of vowel variations from French, a couple Cyrillic characters for Russian, and at least one character that Google tells me was Arabic.
* These suggest that the training corpus skews towards English language data, but there are other languages present.
* How are these tokens picked?

In [ ]:
print(test_text)

Article I
Section 1

All legislative Powers herein granted shall be vested in a Congress of the United States, which shall consist of a Senate and House of Representatives.

Section 2

The House of Representatives shall be composed of Members chosen every second Year by the People of the several States, and the Electors in each State shall have the Qualifications requisite for Electors of the most numerous Branch of the State Legislature.

No Person shall be a Representative who shall not have attained to the Age of twenty five Years, and been seven Years a Citizen of the United States, and who shall not, when elected, be an Inhabitant of that State in which he shall be chosen.

Representatives and direct Taxes shall be apportioned among the several States which may be included within this Union, according to their respective Numbers, which shall be determined by adding to the whole Number of free Persons, including those bound to Service for a Term of Years, and excluding Indians not 

Script:
* This text is from the first article of the US Constitution.
* It's long enough that we can show you the benefits of byte pair encoding.
* The language is somewhat formulaic, and you can see several lines here starting with capital No.
* Earlier, there were several lines starting with capital To.
* So there is a good amount of repetition to leverage in this text.
* I'll show you now how to construct the byte pair encoding.

In [ ]:
text_tokens = list(test_text)
text_tokens[:10]

['A', 'r', 't', 'i', 'c', 'l', 'e', ' ', 'I', '\n']

Script:
* First, I split the text into individual characters as the starting tokens.
* Technically, those should be individual bytes, but they are equivalent for this text.
* Next, lets start counting how many time pairs of adjacent occur.

In [ ]:
def count_pairs():
    counts = {}
    for i in range(len(text_tokens)-1):
        pair = (text_tokens[i], text_tokens[i+1])
        counts[pair] = counts.get(pair, 0) + 1

    return counts

In [ ]:
count_pairs()

{('A', 'r'): 6,
 ('r', 't'): 29,
 ('t', 'i'): 123,
 ('i', 'c'): 48,
 ('c', 'l'): 9,
 ('l', 'e'): 42,
 ('e', ' '): 466,
 (' ', 'I'): 25,
 ('I', '\n'): 1,
 ('\n', 'S'): 10,
 ('S', 'e'): 37,
 ('e', 'c'): 68,
 ('c', 't'): 50,
 ('i', 'o'): 71,
 ('o', 'n'): 165,
 ('n', ' '): 156,
 (' ', '1'): 2,
 ('1', '\n'): 1,
 ('\n', '\n'): 62,
 ('\n', 'A'): 2,
 ('A', 'l'): 3,
 ('l', 'l'): 135,
 ('l', ' '): 132,
 (' ', 'l'): 16,
 ('e', 'g'): 17,
 ('g', 'i'): 13,
 ('i', 's'): 44,
 ('s', 'l'): 9,
 ('l', 'a'): 39,
 ('a', 't'): 157,
 ('i', 'v'): 35,
 ('v', 'e'): 77,
 (' ', 'P'): 59,
 ('P', 'o'): 9,
 ('o', 'w'): 10,
 ('w', 'e'): 8,
 ('e', 'r'): 151,
 ('r', 's'): 50,
 ('s', ' '): 166,
 (' ', 'h'): 39,
 ('h', 'e'): 256,
 ('r', 'e'): 147,
 ('e', 'i'): 29,
 ('i', 'n'): 142,
 (' ', 'g'): 10,
 ('g', 'r'): 23,
 ('r', 'a'): 27,
 ('a', 'n'): 167,
 ('n', 't'): 98,
 ('t', 'e'): 150,
 ('e', 'd'): 98,
 ('d', ' '): 194,
 (' ', 's'): 151,
 ('s', 'h'): 108,
 ('h', 'a'): 136,
 ('a', 'l'): 140,
 (' ', 'b'): 123,
 ('b', 'e'): 99

Script:
* Some of those pairs occur many times and some occur just once.
* We will identify the pairs that occur the most and combine them to form new tokens.

In [ ]:
def get_max_pair():
    counts = count_pairs()
    return max(counts.keys(), key=counts.get)

get_max_pair()

('e', ' ')

Script:
* The most common pair of characters is the letter e followed by a space.
* This pair happens for any word ending in e that is not at the end of the sentence or line.

In [ ]:
def add_new_token():
    pair = get_max_pair()
    new_token = ''.join(pair)
    print("OLD", repr(pair[0]), " AND ", repr(pair[1]), " MERGED INTO NEW TOKEN", repr(new_token))

    old_length = len(text_tokens)
    i = 0
    while i + 1 < len(text_tokens):
        if tuple(text_tokens[i:i+2]) == pair:
            text_tokens[i:i+2] = [new_token]

        i += 1

    print("  SHRANK FROM", old_length, "TO", len(text_tokens), "TOKENS")

add_new_token()

OLD 'e'  AND  ' '  MERGED INTO NEW TOKEN 'e '
  SHRANK FROM 13234 TO 12768 TOKENS


Script:
* This code looks for all the occurences of this pair in our token list and replaces them with a new token combining them.
* Let's repeat this some more.

In [ ]:
for i in range(10):
    add_new_token()

OLD 't'  AND  'h'  MERGED INTO NEW TOKEN 'th'
  SHRANK FROM 12768 TO 12478 TOKENS
OLD ' '  AND  'th'  MERGED INTO NEW TOKEN ' th'
  SHRANK FROM 12478 TO 12265 TOKENS
OLD ' '  AND  'a'  MERGED INTO NEW TOKEN ' a'
  SHRANK FROM 12265 TO 12076 TOKENS
OLD 'e'  AND  's'  MERGED INTO NEW TOKEN 'es'
  SHRANK FROM 12076 TO 11901 TOKENS
OLD ' '  AND  'o'  MERGED INTO NEW TOKEN ' o'
  SHRANK FROM 11901 TO 11726 TOKENS
OLD ' th'  AND  'e '  MERGED INTO NEW TOKEN ' the '
  SHRANK FROM 11726 TO 11564 TOKENS
OLD 'd'  AND  ' '  MERGED INTO NEW TOKEN 'd '
  SHRANK FROM 11564 TO 11408 TOKENS
OLD 'e'  AND  'n'  MERGED INTO NEW TOKEN 'en'
  SHRANK FROM 11408 TO 11256 TOKENS
OLD 'e'  AND  'r'  MERGED INTO NEW TOKEN 'er'
  SHRANK FROM 11256 TO 11105 TOKENS
OLD 'a'  AND  't'  MERGED INTO NEW TOKEN 'at'
  SHRANK FROM 11105 TO 10955 TOKENS


Script:
* You can see that the token creation process has already discovered the word the and created a token with "the" surrounded by spaces.
* Generally, this process will repeat until a target number of tokens is reached or the new tokens are not occuring very often so they won't add much to the model.

Script: (faculty on screen)
* Tokenization has turned into an important part of how large language models work.
* Longer tokens have the benefits of faster text generation and better quality.
* Byte pair encoding naturally picks the longer tokens that will help a language model the most.